src/compiler/checker.ts

```diff
@@ -24527,17 +24527,25 @@ namespace ts {
             checkGrammarStatementInAmbientContext(node);
 
             checkSourceElement(node.statement);
-            checkExpression(node.expression);
+            checkTruthinessExpression(node.expression);
         }
 
         function checkWhileStatement(node: WhileStatement) {
             // Grammar checking
             checkGrammarStatementInAmbientContext(node);
 
-            checkExpression(node.expression);
+            checkTruthinessExpression(node.expression);
             checkSourceElement(node.statement);
         }
 
+        function checkTruthinessExpression(node: Expression) {
+            const type = checkExpression(node);
+            if (type === voidType) {

```

Comment:
A better way to write this is `if (type.flags & TypeFlags.Void)` as we may at some point introduce other type instances that represent a `void` type.